stumpy/search.py

```diff
@@ -0,0 +1,361 @@
+import numpy as np
+
+from . import core
+
+
+def search_k_motifs(
```

Comment:
I haven't thought this through but are motifs/discords only available for self-joins and not AB-joins?

Reply:
What are motifs between two series supposed to be?

Reply:
Hahahaha - I have no idea. I did preface my question with "I haven't thought this through..." 😄

Reply:
I will think about this while eating my lunch

Reply:
I just learned about Matrix Profiling today, so please ignore this comment if it is completely irrelevant. Just to add to the discussion here;

A motif between two (different - thus AB joined) time series would be something similar to 'finding an similar looking pattern, occurring in two different time series'. 
That would be highly relevant in various fields, such as (but certainly not limited to) predictive maintenance. Let's say two pressure valves both have an IOT sensor with highly different readings under normal conditions. It may be, that both valves show the same pattern right before breaking. Thus, finding the occurrence of a motif extracted from series A within series B may be an entirely valid use case.

Reply:
@RookieRakkert Thanks for chiming in and welcome to STUMPY! So, we do currently allow matrix profiles to be computed between two different time series. Just talking out lout here but, for some time series, `T_B` that contains query subsequences, `T_B[i : i + m]`, the matrix profile would tell you the shortest distance to some other subsequence in time series `T_A[j : j +m]`. Then, the motifs would be the local minima in the matrix profile. For each minima, we'd use the query subsequence `T_B[i : i + m]` to then re-scan `T_A` in search of possible multiple matches (using `mass`). Once complete, we'd move on to the next query subsequence and repeat the process until we have `k` motifs discovered. 

Perhaps, to some extent, we are slowly touching upon the meaning behind computing the `mpdist` for two different time series. But I am diverging.

So, this gives us the `k` subsequences in `T_B` that have reasonably good matches found in `T_A`. What does this mean? Why might it be useful? I'm not entirely sure of all of the use cases/applications.

> What are motifs between two series supposed to be?

@mexxexx Would you mind being more specific with your question?  Are you saying that there is little/no meaning in discovering motifs between two time series? Or are you curious about the actual interpretation of what a motif between two time series means? Or neither?

Reply:
In my understanding a motif is a recurring pattern in a time series. If you compute the MP of two time series, than low MP values indicate that a subsequence in T_A has a close match in T_B. Is this a motif? If I think about it like this, the answer would probably be yes... 

But how would motif discovery look like? With one timeseries the answer is simple: If it has a low MP value, it can be found somewhere else in the time series. Then the motif discovery would just locate all other occurrences. What about two time series? Would it mean the following?: Low MP value => a subsequence Q from T_A has a close match in T_B. The the motif discovery tool would find all other close matches in T_B?

Reply:
> In my understanding a motif is a recurring pattern in a time series. If you compute the MP of two time series, than low MP values indicate that a subsequence in T_A has a close match in T_B. Is this a motif? If I think about it like this, the answer would probably be yes...

Yes, this is how I understand it as well.

> But how would motif discovery look like? With one timeseries the answer is simple: If a subsequence has a low MP value, it can be found somewhere else in the time series. Then the motif discovery would just locate all other occurrences. What about two time series? Would it mean the following: Low MP value => the subsequence from T_A has a close match in T_B. The the motif discovery tool would find all other close matches in T_B?

Yes, I agree. The motif discovery tool would do exactly this. I don't see anything fundamentally or conceptually wrong. Do you?

Reply:
@seanlaw Thanks for your reply. I should've read more into this thread before 

> So, this gives us the `k` subsequences in `T_B` that have reasonably good matches found in `T_A`. What does this mean? Why might it be useful? I'm not entirely sure of all of the use cases/applications.

This is exactly what I tried to answer in my original post. Maybe I didn't express myself correctly, let me try again:

Let's say I've got a sensor dataset with labeled anomalies, `T_A`. We know (due to the label), and indeed find that `T_A` has a discord at location `T_A[i :  i+m]`. Another discord occurred at a different location `T_A[j : j+m]`. With the motif discovery, we have found a motif that occurs at both `T_A[i-z : i-z+m]` and `T_A[j-z : j-z+m]`, with `z` arbitrary chosen just to clarify the example. Thus, my hypothesis is that the subsequence (`T_A[i-z : i-z+m]`) was an _indicator_ of the discord about to happen. 
See this as a pattern leading up to an anomalous event. 

With that in mind, it would be highly valuable to find subsequences in `T_B` that show a reasonable good match with that exact subsequence from `T_A`. It potentially be an indicator that the following (yet to be received data) will become a discord, and take appropriate action before the event happens.

Reply:
@RooieRakkert Yes, I think we understand. Technically, the [`mass`](https://github.com/TDAmeritrade/stumpy/blob/bf80f7c128463d7480fc2513f08187a34e96b53d/stumpy/core.py#L602) function does can take a query, `Q`, (that could either come from `T_A` or `T_B` or any time series) and it computes the distance profile for it relative to any time series that, `T`, that you provide it (again, `T` can be `T_A`, `T_B`, or any other time series). So, if you already know `Q` (regardless of whether or not you know if `Q` is a good match to anything) then you can use `mass` directly  to check if a "match" exists. Here, you don't need the matrix profile.

What @mexxexx is trying to accomplish also leverages `mass` but it's actually a two step process so there's a subtle difference. First, in the case of motifs, he identifies the subsequences that already have a close match as indicated by the matrix profile. Afterward, for each individual subsequence motif, he'll identify locations that are within some threshold. This second step is the same as above but the first (iterative) step requires some additional considerations that we're trying to thoughtfully traverse and ensure that we understand/communicate the terminology to avoid future ambiguities.

Reply:
To me, this sounds like a somewhat different problem. When computing the MP between two time series, you are looking for similarities between `T_A` and `T_B`. But from your description, I understand that `T_A` is something like your training data, and `T_B` would be new data you want to apply a model on.

When speaking about sensors, the approach I described would be "Are there events in the sensor A data that look exactly the same way in sensor B data?", vs you described "From sensor A I learn that pattern C leads to an anomaly. Can I find pattern C in the data of sensor B?". I'm not sure if the difference is clear, its just that to find pattern C you don't need the data of sensor B (that's how I understand your example).

If I did indeed understand your example correctly, you could try the following thing.

1. Find the motifs you are interesting using a self join MP of time series `T_A`. That is calculate  the matrixprofile of `T_A` with itself (i.e. using `stumpy.stump(T_A, m, self_join=True)`), and then search for the motifs the way it will be implemented through this PR.
2. Lets say you found one motif, `Q`. For the data of sensor B, you would then compute the *distance* profile (`core.mass`) between `Q` and `T_B` (you could even do this in a streaming fashion, by computing the distance between `Q` and the `m` newest values of sensor B). If the distance is below a threshold, you would perform some action, e.g. performing a warning.